[DependencyInjection] fixed ini file values conversion

fabpot opened the pull request:
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no-ish |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | n/a |

When using the ini format to load parameters in the Container, the parameter values were converted by PHP directly (`'true'` => `1` for instance). But when using the YAML or XML format, the conversions are much broader and more precise (`'true'` => `true` for instance). This PR fixed fixes this discrepancy by using the same rules as XML (we could use `INI_SCANNER_TYPED` for recent versions of PHP but the rules are not exactly the same, so I prefer consistency here).

One might argue that this is a new feature and that this should be merged into master, which I can accept as well. In master, the `XmlUtils::phpize()` method should be deprecated and replaced by a more generic phpize class.

ping @symfony/deciders

fabpot committed:
[DependencyInjection] fixed ini file values conversion

javiereguiluz commented:
👍 

It can be debatable to merge it in 2.7 or master ... but to me this is a bugfix, so it should stay in 2.7.

dunglas commented:
👍

nicolas-grekas commented:
:+1: for 2.7 also

ro0NL commented:
@fabpot can you test with `on`, `off`, `yes`, `no`... i think this can get problematic ;-) The BC break is real: https://3v4l.org/LuoJt

fabpot commented:
@ro0NL Not sure I understand your concern. Can you be more precise?

ro0NL commented:
This breaks configuration like
```ini
[parameters]
   some_key = on
```

right? AFAIK this is not covered by `XmlUtils::phpize` nor this PR...

ro0NL commented:
@fabpot perhaps more serious, no escaping happens: https://3v4l.org/0E5k7

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Loader/IniFileLoader.php:
```diff
@@ -30,14 +31,14 @@ public function load($resource, $type = null)
         $this->container->addResource(new FileResource($path));
 
-        $result = parse_ini_file($path, true);
+        $result = parse_ini_file($path, true, INI_SCANNER_RAW);
```
wouldn't work, nevermind...

nicolas-grekas commented:
@ro0NL is right, we need to polyfill `INI_SCANNER_TYPED`...

ro0NL commented:
@nicolas-grekas we basically already doing that with `INI_SCANNER_NORMAL + XmlUtils::phpize`.

Imo a

> more generic phpize class

would be sufficient.

ro0NL commented on src/Symfony/Component/DependencyInjection/Loader/IniFileLoader.php:
```diff
@@ -49,4 +50,32 @@ public function supports($resource, $type = null)
+                ("'" === $value[0] && "'" === $value[strlen($value) - 1]) ||
+                ('"' === $value[0] && '"' === $value[strlen($value) - 1])
+            ):
+                // quoted string
```
Perhaps throw if `1 === strlen($value)` (invalid syntax)?

fabpot replied:
the php.ini file would not be valid in that case.

ro0NL replied:
It was fatal before :/ https://3v4l.org/VgDvJ

With SCANNER_RAW it parses though.. but that doesnt mean it's valid, it's raw :)

fabpot replied:
fixed

fabpot commented:
I've updated this PR to take into account how the scanner type convert values, and added some more tests. As for the failure on PHP 7.1, not sure what changed. Is it better now?

fabpot commented:
I propose to merge it into master instead of 2.7 as it's more complex than anticipated.

nicolas-grekas commented:
Using `INI_SCANNER_TYPED` when possible would be a great way to ensure the polyfill behaves as expected

fabpot commented:
@nicolas-grekas Not really possible as our behavior is not exactly the same.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Loader/IniFileLoader.php:
```diff
@@ -49,4 +54,32 @@ public function supports($resource, $type = null)
+                return true;
+            case 'no' === $value || 'off' === $value || 'none' === $value:
+                return false;
+            case isset($value[0]) && (
```
I propose this patch:
```diff
--- a/src/Symfony/Component/DependencyInjection/Loader/IniFileLoader.php
+++ b/src/Symfony/Component/DependencyInjection/Loader/IniFileLoader.php
@@ -64,15 +64,16 @@ class IniFileLoader extends FileLoader
    {
        // trim on the right as a comment leaves the whitespaces in place
        $value = rtrim($value);
+        $lowercaseValue = strtolower($value);

        switch (true) {
            case defined($value):
                return constant($value);
-            case 'yes' === $value || 'on' === $value:
+            case 'yes' === $lowercaseValue || 'on' === $lowercaseValue:
                return true;
-            case 'no' === $value || 'off' === $value || 'none' === $value:
+            case 'no' === $lowercaseValue || 'off' === $lowercaseValue || 'none' === $lowercaseValue:
                return false;
-            case isset($value[0]) && (
+            case isset($value[1]) && (
                ("'" === $value[0] && "'" === $value[strlen($value) - 1]) ||
                ('"' === $value[0] && '"' === $value[strlen($value) - 1])
            ):
diff --git a/src/Symfony/Component/DependencyInjection/composer.json b/src/Symfony/Component/DependencyInjection/composer.json
index 9c12d11..e24187f 100644
--- a/src/Symfony/Component/DependencyInjection/composer.json
+++ b/src/Symfony/Component/DependencyInjection/composer.json
@@ -20,7 +20,7 @@
    },
    "require-dev": {
        "symfony/yaml": "~2.3.42|~2.7.14|~2.8.7",
-        "symfony/config": "~2.2",
+        "symfony/config": "~2.7",
        "symfony/expression-language": "~2.6"
    },
    "conflict": {
```

fabpot commented:
Rebased on master
